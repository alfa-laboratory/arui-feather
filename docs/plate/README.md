# Plate

Компонент плашки.

```javascript
import Plate from 'arui-feather/plate';
```




## Props


| Prop  | Тип  | По-умолчанию | Обязательный | Описание |
| ----- | ---- | ------------ | ------------ |----------|
| hasCloser | Boolean | `false`  |  | Управление наличием закрывающего крестика |
| isFlat | Boolean |  |  | Плоская тема |
| children | Array.<Node>\|Node |  |  | Дочерние элементы `Plate` |
| theme | [ThemeEnum](#ThemeEnum) |  |  | Тема компонента |
| className | Function\|String |  |  | Дополнительный класс |
| onClick | Function |  |  | Обработчик клика по плашке |
| onCloserClick | Function |  |  | Обработчик клика по крестику |







## Типы






### <a id="ThemeEnum"></a>ThemeEnum

 * `'alfa-on-color'`
 * `'alfa-on-white'`



