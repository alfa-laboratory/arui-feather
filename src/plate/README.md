```jsx
<Plate
    hasCloser={ true }
>
    <Heading size='s'>
        Альфа-Мобайл 9.0
    </Heading>
    <Paragraph>
        Новый, превосходный, невероятный, беспрецедентный, большой, внушительный, четкий, безотказный, волшебный, понятный. +40 новых возможностей твоего мобайла!
    </Paragraph>
</Plate>
```
