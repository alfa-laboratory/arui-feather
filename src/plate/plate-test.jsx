/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { render, cleanUp } from '../test-utils';

import Plate from './plate';

describe('plate', () => {
    afterEach(cleanUp);

    it('should render without problems', () => {
        let plate = render(<Plate>plate-text</Plate>);

        expect(plate.node).to.exist;
        expect(plate.node).to.have.text('plate-text');
        expect(plate.node).to.have.class('plate');
    });

    it('should render with cross without problems', () => {
        let plate = render(<Plate hasCloser={ true }>plate-text</Plate>);
        let crossNode = plate.node.querySelector('.plate__closer');

        expect(crossNode).to.exist;
    });

    it('should call `onCloserClick` callback after plate closer was clicked', () => {
        let onCloserClick = sinon.spy();
        let plate = render(
            <Plate hasCloser={ true } onCloserClick={ onCloserClick }>plate-text</Plate>
        );
        let closer = plate.node.querySelector('.plate__closer');

        closer.click();

        expect(onCloserClick).to.have.been.calledOnce;
    });

    it('should call `onClick` callback after plate was clicked', () => {
        let onClick = sinon.spy();
        let plate = render(
            <Plate hasCloser={ true } onClick={ onClick }>plate-text</Plate>
        );

        plate.node.click();

        expect(onClick).to.have.been.calledOnce;
    });
});
