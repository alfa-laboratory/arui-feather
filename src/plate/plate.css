/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

@import '../vars.css';

.plate {
    display: inline-block;
    position: relative;
    margin-right: 7px;
    margin-bottom: 7px;
    padding: var(--gap-s);
    border-radius: var(--border-radius-control);
    box-shadow: var(--shadow);
    background-color: #fff;

    &_hidden {
        display: none;
    }

    &_has-closer {
        padding-right: 34px;
    }

    &__closer {
        opacity: var(--opacity-minor);
        position: absolute;
        top: 0;
        right: 0;
        cursor: pointer;
        transition: opacity .1s ease-out;

        &:hover {
            opacity: 1;
        }
    }

    &_flat {
        box-shadow: none;
    }

    &_flat& {
        &_theme_alfa-on-color {
            background-color: var(--color-background-minor-alfa-on-color);
        }

        &_theme_alfa-on-white {
            background-color: var(--color-background-minor-alfa-on-white);
        }
    }

    &__content {
        color: var(--color-content-accent-alfa-on-white);
    }

    &_flat&_theme_alfa-on-color & {
        &__content {
            color: var(--color-content-accent-alfa-on-color);
        }
    }
}
