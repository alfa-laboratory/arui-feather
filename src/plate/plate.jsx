/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/* eslint jsx-a11y/no-static-element-interactions: 0 */

import { autobind } from 'core-decorators';
import React from 'react';
import Type from 'prop-types';

import Icon from '../icon/icon';
import IconButton from '../icon-button/icon-button';

import cn from '../cn';
import performance from '../performance';

/**
 * Компонент плашки.
 */
@cn('plate')
@performance()
class Plate extends React.Component {
    static propTypes = {
        /** Управление наличием закрывающего крестика */
        hasCloser: Type.bool,
        /** Плоская тема */
        isFlat: Type.bool,
        /** Дочерние элементы `Plate` */
        children: Type.oneOfType([Type.arrayOf(Type.node), Type.node]),
        /** Тема компонента */
        theme: Type.oneOf(['alfa-on-color', 'alfa-on-white']),
        /** Дополнительный класс */
        className: Type.string,
        /** Идентификатор компонента в DOM */
        id: Type.string,
        /** Обработчик клика по плашке */
        onClick: Type.func,
        /** Обработчик клика по крестику */
        onCloserClick: Type.func,
        /** Обработчик события нажатия на клавишу клавиатуры в момент, когда фокус находится на компоненте */
        onKeyDown: Type.func
    };

    static defaultProps = {
        hasCloser: false
    };

    state = {
        isHidden: false
    };

    render(cn) {
        let { children } = this.props;

        if (Array.isArray(children)) {
            children = React.Children.map(
                children, child => React.cloneElement(
                    child, { theme: this.getContentTheme() }
                )
            );
        }

        return (
            <div
                className={ cn({
                    'has-closer': this.props.hasCloser,
                    hidden: this.props.hasCloser && this.state.isHidden,
                    flat: this.props.isFlat
                }) }
                id={ this.props.id }
                onClick={ this.handleClick }
                onKeyDown={ this.handleKeyDown }
            >
                <div className={ cn('content') }>
                    { children }
                    {
                        this.props.hasCloser &&
                            <IconButton
                                className={ cn('closer') }
                                onClick={ this.handleCloserClick }
                            >
                                <Icon
                                    theme={ this.getContentTheme() }
                                    name='tool-close'
                                />
                            </IconButton>
                    }
                </div>
            </div>
        );
    }

    @autobind
    handleClick(event) {
        if (this.props.onClick) {
            this.props.onClick(event);
        }
    }

    @autobind
    handleCloserClick() {
        this.setState({
            isHidden: true
        });

        if (this.props.onCloserClick) {
            this.props.onCloserClick();
        }
    }

    @autobind
    handleKeyDown(event) {
        if (this.props.onKeyDown) {
            this.props.onKeyDown(event);
        }
    }

    @autobind
    getContentTheme() {
        return this.props.isFlat ? this.props.theme : 'alfa-on-white';
    }
}

export default Plate;
