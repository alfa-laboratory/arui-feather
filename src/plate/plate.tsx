/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/* eslint jsx-a11y/no-static-element-interactions: 0 */

import React from 'react';
import { createCn } from 'bem-react-classname';

import IconClose from '../icon/ui/close';
import IconArrowUp from '../icon/ui/arrow-up';
import IconArrowDown from '../icon/ui/arrow-down';
import IconButton from '../icon-button/themed';

export type PlateProps = {

    /**
     * Управление наличием закрывающего крестика
     */
    hasCloser?: boolean;

    /**
     * Управление наличием стрелки скрытия контента
     */
    foldable?: boolean;

    /**
     * Начальное состояние контента при foldable={ true }
     */
    folded?: boolean;

    /**
     * Иконка компонента
     */
    icon?: React.ReactNode;

    /**
     * Дочерние элементы `Plate`
     */
    children?: React.ReactNode;

    /**
     * Тема компонента
     */
    theme?: 'alfa-on-color' | 'alfa-on-white';

    /**
     * Заголовок `Plate`
     */
    title?: string;

    /**
     * Тип `Plate`
     */
    type?: 'default' | 'common' | 'error' | 'success';

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Идентификатор компонента в DOM
     */
    id?: string;

    /**
     * Обработчик клика стрелке сворачивания\разворачивания плашки
     */
    onFolderClick?: (event?: React.MouseEvent<any>) => void;

    /**
     * Обработчик клика на заголовке
     */
    onTitleClick?: (event?: React.MouseEvent<any>) => void;

    /**
     * Обработчик события нажатия на клавишу клавиатуры в момент, когда фокус находится на заголовке компонента
     */
    onTitleKeyDown?: (event?: React.KeyboardEvent<any>) => void;

    /**
     * Обработчик клика по плашке
     */
    onClick?: (event?: React.MouseEvent<any>) => void;

    /**
     * Обработчик клика по крестику
     */
    onCloserClick?: (event?: React.MouseEvent<any>) => void;

    /**
     * Обработчик события нажатия на клавишу клавиатуры в момент, когда фокус находится на компоненте
     */
    onKeyDown?: (event?: React.KeyboardEvent<any>) => void;

    /**
     * Идентификатор для систем автоматизированного тестирования
     */
    'data-test-id'?: string;

};

/**
 * Компонент плашки.
 */
class Plate extends React.PureComponent<PlateProps> {
    cn = createCn('plate');

    static defaultProps: Partial<PlateProps> = {
        foldable: false,
        folded: true,
        type: 'default',
        title: ''
    };

    state = {
        isHidden: false,
        isFolded: this.props.folded
    };

    root: HTMLElement;

    render() {
        return (
            <div
                className={ this.cn({
                    'has-closer': this.props.hasCloser,
                    foldable: this.props.foldable,
                    'has-icon': !!this.props.icon,
                    hidden: this.props.hasCloser && this.state.isHidden,
                    type: this.props.type
                }) }
                id={ this.props.id }
                onClick={ this.handleClick }
                onKeyDown={ this.handleKeyDown }
                ref={ (node) => {
                    this.root = node;
                } }
                data-test-id={ this.props['data-test-id'] }
            >
                {
                    this.props.icon &&
                    <span className={ this.cn('icon') }>
                        { this.props.icon }
                    </span>
                }
                {
                    this.props.title &&
                    <div
                        className={ this.cn('title') }
                    >
                        <span
                            onClick={ this.handleTitleClick }
                            onKeyUp={ this.handleTitleKeyDown }
                        >
                            { this.props.title }
                        </span>
                        {
                            this.props.foldable &&
                            <IconButton
                                className={ this.cn('folder') }
                                onClick={ this.handleFolderClick }
                            >
                                {
                                    this.state.isFolded
                                        ? <IconArrowDown />
                                        : <IconArrowUp />
                                }
                            </IconButton>
                        }
                    </div>
                }
                <div
                    className={ this.cn('content', {
                        folded: this.props.foldable && this.state.isFolded
                    }) }
                >
                    { this.props.children }
                    {
                        this.props.foldable ||
                        (this.props.hasCloser &&
                        <IconButton
                            className={ this.cn('closer') }
                            onClick={ this.handleCloserClick }
                        >
                            <IconClose />
                        </IconButton>)
                    }
                </div>
            </div>
        );
    }

    handleClick = (event) => {
        if (this.props.onClick) {
            this.props.onClick(event);
        }
    };

    handleTitleClick = (event) => {
        if (this.props.foldable) {
            this.setState({
                isFolded: !this.state.isFolded
            });
        }

        if (this.props.onTitleClick) {
            this.props.onTitleClick(event);
        }
    }

    handleTitleKeyDown = (event) => {
        if (this.props.onTitleKeyDown) {
            this.props.onTitleKeyDown(event);
        }
    }

    handleFolderClick = (event) => {
        this.setState({
            isFolded: !this.state.isFolded
        });

        if (this.props.onFolderClick) {
            this.props.onFolderClick(event);
        }
    }

    handleCloserClick = (event) => {
        this.setState({
            isHidden: true
        });

        if (this.props.onCloserClick) {
            this.props.onCloserClick(event);
        }
    };

    handleKeyDown = (event) => {
        if (this.props.onKeyDown) {
            this.props.onKeyDown(event);
        }
    }
}

export default Plate;
