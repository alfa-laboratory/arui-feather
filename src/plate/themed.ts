import { withTheme } from '../cn';
import Component from './plate';

export default withTheme(Component);
