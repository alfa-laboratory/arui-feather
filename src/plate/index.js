import '../icon';

import './plate.css';

export default from './plate';
