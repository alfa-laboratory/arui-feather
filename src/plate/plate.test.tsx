/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import React from 'react';
import { shallow, mount } from 'enzyme';

import { Plate } from './plate';
import IconOk from '../icon/ui/ok';

describe('plate', () => {
    it('should render without problems', () => {
        const plate = shallow(<Plate>plate-text</Plate>);

        expect(plate).toMatchSnapshot();
        expect(plate.html()).toContain('plate-text');
    });

    it('should render with cross without problems', () => {
        const plate = shallow(<Plate hasCloser={ true }>plate-text</Plate>);
        const crossNode = plate.find('.plate__closer');

        expect(crossNode).toBeTruthy();
    });

    it('should call `onCloserClick` callback after plate closer was clicked', () => {
        const onCloserClick = jest.fn();
        const plate = shallow(
            <Plate hasCloser={ true } onCloserClick={ onCloserClick }>plate-text</Plate>
        );
        const closer = plate.find('.plate__closer');

        closer.simulate('click');

        expect(onCloserClick).toHaveBeenCalled();
    });

    it('should call `onClick` callback after plate was clicked', () => {
        const onClick = jest.fn();
        const plate = mount(
            <Plate hasCloser={ true } onClick={ onClick }>plate-text</Plate>
        );

        plate.simulate('click');

        expect(onClick).toHaveBeenCalled();
    });

    it('should call `onFolderClick` callback after plate folder was clicked', () => {
        const onFolderClick = jest.fn();
        const plate = shallow(
            <Plate title='plate-title' foldable={ true } onFolderClick={ onFolderClick }>plate-text</Plate>
        );
        const folder = plate.find('.plate__folder');

        folder.simulate('click');

        expect(onFolderClick).toHaveBeenCalled();
    });

    it('should call `onTitleClick` callback after plate folder was clicked', () => {
        const onTitleClick = jest.fn();
        const plate = shallow(
            <Plate title='plate-title' foldable={ true } onTitleClick={ onTitleClick }>plate-text</Plate>
        );
        const title = plate.find('span');

        title.simulate('click');

        expect(onTitleClick).toHaveBeenCalled();
    });

    it('should render icon when it propperty is setted', () => {
        const plate = shallow(
            <Plate
                title='Успешный'
                type='success'
                hasCloser={ true }
                icon={
                    <IconOk
                        colored={ true }
                    />
                }
            />
        );

        expect(plate).toMatchSnapshot();
        expect(plate.html()).toContain('plate__icon');
    });

});
