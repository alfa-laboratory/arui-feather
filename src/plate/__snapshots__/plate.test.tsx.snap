// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plate should render icon when it propperty is setted 1`] = `
<div
  className="plate plate_has-closer plate_has-icon plate_type_success"
  onClick={[Function]}
  onKeyDown={[Function]}
>
  <span
    className="plate__icon"
  >
    <ForwardRef(ThemedComponent(IconOk))
      colored={true}
    />
  </span>
  <div
    className="plate__title"
  >
    <span
      onClick={[Function]}
      onKeyUp={[Function]}
    >
      Успешный
    </span>
  </div>
  <div
    className="plate__content"
  >
    <ForwardRef(ThemedComponent(IconButton))
      className="plate__closer"
      onClick={[Function]}
    >
      <ForwardRef(ThemedComponent(IconClose)) />
    </ForwardRef(ThemedComponent(IconButton))>
  </div>
</div>
`;

exports[`plate should render without problems 1`] = `
<div
  className="plate plate_type_default"
  onClick={[Function]}
  onKeyDown={[Function]}
>
  <div
    className="plate__content"
  >
    plate-text
  </div>
</div>
`;
